source/sundials/arkode.cc

```diff
@@ -499,20 +1096,167 @@ namespace SUNDIALS
+    , A_data(A_data)
+    , a_times_fn(a_times_fn)
+
+  {
+    Assert(a_times_fn != nullptr, ExcInternalError());
+  }
+
+
+
+  template <typename VectorType>
+  void
+  SundialsOperator<VectorType>::vmult(VectorType &      dst,
+                                      const VectorType &src) const
+  {
+    N_Vector sun_dst = solver.create_vector(dst);
+    N_Vector sun_src = solver.create_vector(src);
+    copy(sun_src, src);
+    int status = a_times_fn(A_data, sun_src, sun_dst);
+    (void)status;
+    AssertARKode(status);
+
+    copy(dst, sun_dst);
+
+    solver.free_vector(sun_dst);
+    solver.free_vector(sun_src);
+  }
+
+
+
+  template <typename VectorType>
+  SundialsPreconditioner<VectorType>::SundialsPreconditioner(
+    ARKode<VectorType> &solver,
+    void *              P_data,
+    PSolveFn            p_solve_fn,
+    double              tol)
+    : solver(solver)
+    , P_data(P_data)
+    , p_solve_fn(p_solve_fn)
+    , tol(tol)
+  {}
+
+
+
+  template <typename VectorType>
+  void
+  SundialsPreconditioner<VectorType>::vmult(VectorType &      dst,
+                                            const VectorType &src) const
+  {
+    // apply identity preconditioner if nothing else specified
+    if (!p_solve_fn)
+      {
+        dst = src;
+        return;
+      }
+
+    N_Vector sun_dst = solver.create_vector(dst);
+    N_Vector sun_src = solver.create_vector(src);
+    copy(sun_src, src);
+    // for custom preconditioners no distinction between left and right
+    // preconditioning is made
+    int status =
+      p_solve_fn(P_data, sun_src, sun_dst, tol, 0 /*precondition_type*/);
+    (void)status;
+    AssertARKode(status);
+
+    copy(dst, sun_dst);
+
+    solver.free_vector(sun_dst);
+    solver.free_vector(sun_src);
+  }
+#  endif
+
   template class ARKode<Vector<double>>;
   template class ARKode<BlockVector<double>>;
 
+#  if DEAL_II_SUNDIALS_VERSION_GTE(4, 0, 0)
+  template struct SundialsOperator<Vector<double>>;
+  template struct SundialsOperator<BlockVector<double>>;
+
+  template struct SundialsPreconditioner<Vector<double>>;
+  template struct SundialsPreconditioner<BlockVector<double>>;
+#  endif
+
 #  ifdef DEAL_II_WITH_MPI
 
 #    ifdef DEAL_II_WITH_TRILINOS
   template class ARKode<TrilinosWrappers::MPI::Vector>;
   template class ARKode<TrilinosWrappers::MPI::BlockVector>;
+
+#      if DEAL_II_SUNDIALS_VERSION_GTE(4, 0, 0)
+  template struct SundialsOperator<TrilinosWrappers::MPI::Vector>;
+  template struct SundialsOperator<TrilinosWrappers::MPI::BlockVector>;
+
+  template struct SundialsPreconditioner<TrilinosWrappers::MPI::Vector>;
+  template struct SundialsPreconditioner<TrilinosWrappers::MPI::BlockVector>;
+#      endif
 #    endif // DEAL_II_WITH_TRILINOS
 
 #    ifdef DEAL_II_WITH_PETSC
 #      ifndef PETSC_USE_COMPLEX
   template class ARKode<PETScWrappers::MPI::Vector>;
   template class ARKode<PETScWrappers::MPI::BlockVector>;
+
+  template class SundialsOperator<PETScWrappers::MPI::Vector>;
```

Comment:
can you please wrap this statement with the next 4 lines into an appropriate `ifdef`, i.e., 
```suggestion

#      if DEAL_II_SUNDIALS_VERSION_GTE(4, 0, 0)
  template class SundialsOperator<PETScWrappers::MPI::Vector>;
```
like the other cases with Trilinos and deal.II vectors? This should fix the CI error.